src/xwayland/xserver.rs

```diff
@@ -1,509 +1,382 @@
+                // to xwayland.
+                unset_cloexec(wayland_socket_fd)?;
+                unset_cloexec(wm_socket_fd)?;
+                unset_cloexec(pipe_fd)?;
+                for &socket in socket_fds.iter() {
+                    unset_cloexec(socket)?;
+                }
+
+                Ok(())
+            });
         }
-    }
-}
 
-impl XWaylandClientData {
-    /// Access user_data map for a xwayland client
-    pub fn user_data(&self) -> &UserDataMap {
-        &self.data_map
+        info!("spawning XWayland instance");
+
+        let _ = command.spawn()?;
+
+        // SAFETY: RawFd's AsRawFd impl is infallible.
+        let wrapper = unsafe { calloop::generic::FdWrapper::new(displayfd_recv.as_raw_fd()) };
+        let source = calloop::generic::Generic::new(wrapper, calloop::Interest::READ, calloop::Mode::Level);
+        let inner = Instance {
+            display_lock: lock,
+            display_fd: displayfd_recv,
+            x11_socket: Some(x_wm_me),
+        };
+
+        let data_map = UserDataMap::new();
+        user_data(&data_map);
+
+        // Insert the client into the display handle. The order is important
+        // here; XWayland never starts up at all unless it can roundtrip with
+        // wayland.
+        let inner = Arc::new(Mutex::new(inner));
+        let mut dh = dh.clone();
+        let client = dh.insert_client(
+            wl_me,
+            Arc::new(XWaylandClientData {
+                #[cfg(feature = "wayland_frontend")]
+                compositor_state: CompositorClientState::default(),
+                data_map,
+            }),
+        )?;
+
+        Ok((
+            Self {
+                inner,
+                source,
+                dh,
+                client: client.clone(),
+            },
+            client,
+        ))
     }
-}
 
-// Launch an XWayland server
-//
-// Does nothing if there is already a launched instance
-fn launch<D, K, V, I, F>(
-    inner: &Arc<Mutex<Inner>>,
-    loop_handle: LoopHandle<'_, D>,
-    mut dh: DisplayHandle,
-    display: Option<u32>,
-    envs: I,
-    open_abstract_socket: bool,
-    user_data: F,
-) -> io::Result<u32>
-where
-    I: IntoIterator<Item = (K, V)>,
-    K: AsRef<OsStr>,
-    V: AsRef<OsStr>,
-    F: FnOnce(&UserDataMap),
-{
-    let mut guard = inner.lock().unwrap();
-    if let Some(instance) = guard.instance.as_ref() {
-        return Ok(instance.display_lock.display());
+    /// Returns the X11 display used by the instance, suitable for setting the
+    /// `DISPLAY` environment variable.
+    pub fn display_number(&self) -> u32 {
+        self.inner.lock().unwrap().display_lock.display_number()
     }
 
-    info!("Starting XWayland");
+    /// Returns a file descriptor which can be polled for readiness. When the fd
+    /// is readable, the XWayland server's readiness can be checked with
+    /// [take_socket](Self::take_socket).
+    pub fn poll_fd(&self) -> BorrowedFd<'_> {
```

Comment:
This could just be an `AsFd` implementation?

Reply:
This matches `DisplayHandle`, fwiw.